src/sources/dnstap/tcp.rs

```diff
@@ -0,0 +1,382 @@
+use ipnet::IpNet;
+use std::time::Duration;
+
+use base64::prelude::{Engine as _, BASE64_STANDARD};
+use bytes::Bytes;
+use serde_with::serde_as;
+use vector_lib::configurable::configurable_component;
+use vector_lib::internal_event::{
+    ByteSize, BytesReceived, InternalEventHandle, Protocol, Registered,
+};
+use vector_lib::ipallowlist::IpAllowlistConfig;
+use vector_lib::lookup::{owned_value_path, path};
+use vector_lib::tcp::TcpKeepaliveConfig;
+use vector_lib::tls::{CertificateMetadata, MaybeTlsSettings, TlsSourceConfig};
+use vector_lib::EstimatedJsonEncodedSizeOf;
+use vrl::path::{OwnedValuePath, PathPrefix};
+use vrl::value::ObjectMap;
+
+use crate::internal_events::{DnstapParseError, SocketEventsReceived, SocketMode};
+use crate::sources::util::framestream::{FrameHandler, TcpFrameHandler};
+use crate::{
+    event::{Event, LogEvent},
+    sources::util::net::SocketListenAddr,
+};
+
+use crate::sources::dnstap::parser::DnstapParser;
+use crate::sources::dnstap::schema::DNSTAP_VALUE_PATHS;
+use vector_lib::config::{log_schema, LegacyKey, LogNamespace};
+use vector_lib::lookup::lookup_v2::OptionalValuePath;
+
+/// TCP configuration for the `dnstap` source.
+#[serde_as]
+#[configurable_component]
+#[cfg_attr(unix, serde(tag = "mode", rename_all = "snake_case"))]
+#[derive(Clone, Debug)]
+pub struct TcpConfig {
+    /// Maximum DNSTAP frame length that the source accepts.
+    ///
+    /// If any frame is longer than this, it is discarded.
+    #[serde(default = "default_max_frame_length")]
+    #[configurable(metadata(docs::type_unit = "bytes"))]
+    pub max_frame_length: usize,
+
+    #[configurable(derived)]
+    address: SocketListenAddr,
+
+    #[configurable(derived)]
+    keepalive: Option<TcpKeepaliveConfig>,
+
+    /// The timeout before a connection is forcefully closed during shutdown.
+    #[serde(default = "default_shutdown_timeout_secs")]
+    #[serde_as(as = "serde_with::DurationSeconds<u64>")]
+    #[configurable(metadata(docs::human_name = "Shutdown Timeout"))]
+    shutdown_timeout_secs: Duration,
+
+    /// Overrides the name of the log field used to add the source path to each event.
+    ///
+    /// The value is the socket path itself.
+    ///
+    /// By default, the [global `log_schema.host_key` option][global_host_key] is used.
+    ///
+    /// [global_host_key]: https://vector.dev/docs/reference/configuration/global-options/#log_schema.host_key
+    pub host_key: Option<OptionalValuePath>,
+
+    /// Overrides the name of the log field used to add the peer host's port to each event.
+    ///
+    /// The value will be the peer host's port i.e. `9000`.
+    ///
+    /// By default, `"port"` is used.
+    ///
+    /// Set to `""` to suppress this key.
+    #[serde(default = "default_port_key")]
+    pub port_key: OptionalValuePath,
+
+    /// List of allowed origin IP networks
+    ///
+    /// By default, no origin is allowed
+    permit_origin: IpAllowlistConfig,
```

Comment:
I set it as mandatory, since we discussed above that it would be good to have it secure by default and no value would block all connections. If we don't want that, then I can make it optional, similar to how I added it to socket TCP source.

Reply:
I see. I'm going back and forth between making dnstap more secure or more consistent. I'll check in with others in the team and see if we can come to a consensus on which way we want to go.

Reply:
Hi. Could you make this field optional please? It's more important to be consistent with other components at present. Secure by default is a good idea, but if we want to approach it we should do it throughout Vector rather than just in some areas.